### ISSUE TITLE
Run functions sequentially with same input and environment schemas?

### ISSUE THREAD

**Issue by jam-fran** [seasonedcc/domain-functions#79]
Hi! I've run into a few use cases where I'd like to combine 2 or more domain functions where:

1. They are run sequentially, not in parallel (like the behavior of `pipe` or `sequence`)
2. They are all passed the same input and environment (like the behavior of `and` or `collect`)

I might be missing something or misreading the docs - is this currently possible? If not, any ideas for a workaround or elegant way to handle this? Thank you!

**Comment by gustavoguichard** [comment]
Hi @jam-fran , I just had a scenario where I felt a similar need but ended up solving it with the built-in methods.
I hope it makes sense to you.

### Scenario
I have a `getUser` (`DomainFunction<{ user: User }>`) and a `getUserAssessments` (`DomainFunction<{ assessments: Assessment[] }>`) that takes the user as input and can be used sequentially with a `pipe` operator.
I want to merge their results like this: `DomainFunction<{ user: User, assessments: Assessment[] }>` so I can do:
```ts
const getData = map(
  sequence(getUser, getUserAssessments),
  mergeObjects,
);
```

**All good.**

**Now I have a new requirement that imposes a challenge:** I want to be able to override some properties of `User` on the fly to make different queries for `getUserAsessments` but I do want the full user in my result.

#### Here is my solution:
First I start changing my getUserAssessments' inputSchema:
```ts
// from this
z.object({ user: userSchema })

// to this
z.object({ user: userSchema, overrides: partialUserSchema.optional() })
```

Now all I need to do is to create a `passthrough` domain function:
```ts
const passthrough: makeDomainFunction(z.any())((input) => input)
```

Now I can merge the `passthrough` with `getUser` and send my overrides like so:
```ts
const getData = map(
  sequence(merge(getUser, passthrough), getUserAssessments),
  mergeObjects,
);

// than in my loader:
export async function loader({ request, params }: DataFunctionArgs) {
  const result = await getData({ ...params, overrides: inputFromUrl(request) });
  // ...
}
```

And I can now change the user properties by using a query string: `/my-route?age=20`

### That's it
I hope it helps you finding a nice way to forward your input to sequential domain functions ;)

**Comment by atreib** [comment]
Hey hey! I'd a similar use-case as the two examples above. 

I have a sequence of functions (not parallel), but I need to merge the output of the "previous function" with the input for the "next function", and it should return a single merged output of all functions. 

I needed something like this:

```ts
it("Should aggregate the output on the input of the next function", async () => {
  const input = { a: 1 };

  const dfReturnsB = makeDomainFunction(z.any())(() => ({
    b: 2,
  }));

  const dfReturnsC = makeDomainFunction(
    z.object({ a: z.number(), b: z.number() })
  )(({ a, b }) => ({ a, b, c: 3 }));

  const dfReturnsD = makeDomainFunction(z.any())(() => ({
    d: 4,
  }));

  const aggregatedDomainFunction = aggregate(
    dfReturnsB,
    dfReturnsC,
    dfReturnsD
  );

  const aggregationResult = await fromSuccess(aggregatedDomainFunction)(input);

  expect(aggregationResult).toEqual({
    a: 1,
    b: 2,
    c: 3,
    d: 4,
  });
});
```

And I have ended up with the following function:

```ts
import {
  DomainFunction,
  Result,
  SuccessResult,
  UnpackAll,
  mergeObjects,
} from "domain-functions";

function aggregate<Fns extends DomainFunction[]>(
  ...fns: Fns
): DomainFunction<UnpackAll<Fns>> {
  return async function (initialInput: unknown, environment?: unknown) {
    const results = [];
    let inputs = [initialInput];

    let prevResult: undefined | Result<unknown>;
    for await (const fn of fns as DomainFunction[]) {
      if (prevResult?.success) inputs.push(prevResult.data);
      const result = await fn(mergeObjects(inputs), environment);
      if (!result.success) return result;
      prevResult = result;
      results.push(result.data);
    }

    return {
      success: true,
      data: mergeObjects(results),
      inputErrors: [],
      environmentErrors: [],
      errors: [],
    } as SuccessResult<UnpackAll<Fns>>;
  };
}

export { aggregate };
```

Maybe this approach can help with your cases @jam-fran @gustavoguichard 🙂

Also I would be glad to contribute to the repo, if this function makes sense for the library.
Cheers

**Comment by gustavoguichard** [comment]
I'm reopening this issue as I've been stumbling upon this necessity so I want to investigate if @atreib 's solution could be a good addition to this library.
Andre, I'll circle back to you whenever I test it... would you be willing to send a PR ?

**Comment by diogob** [comment]
couldn't this be built with a sequence + map?

**Comment by atreib** [comment]
@diogob I don't think so 🤔 Because with the `aggregate()`, every step receives the same input + the output of the previous functions. And with the `sequence()` + `map()`, they would all receive the same initial input, right? I'm open to think on other approaches though 🙂 

@gustavoguichard Yeah, 100%!

**Comment by diogob** [comment]
@atreib could you solve your use case moving this initial input to the environment? 

In this way the whole composition would have access to it and you still have the output from the previous function in a `sequence`.
